src/js/media-controller.js

```diff
@@ -38,357 +52,724 @@ const ButtonPressedKeys = ['ArrowLeft', 'ArrowRight', 'Enter', ' ', 'f', 'm', 'k
-    return super.observedAttributes.concat('nohotkeys', 'hotkeys', 'default-stream-type');
-  }
+const MediaUIStates = {
+  MEDIA_PAUSED: {
+    get: function(controller){
+      const { media } = controller;
+
+      return media ? media.paused : true;
+    },
+    mediaEvents: ['play', 'pause', 'emptied']
+  }, 
+  MEDIA_HAS_PLAYED: {
+    // We want to let the user know that the media started playing at any point (`media-has-played`).
+    // Since these propagators are all called when boostrapping state, let's verify this is
+    // a real playing event by checking that 1) there's media and 2) it isn't currently paused.
+    get: function(controller) {
+      const { media } = controller;
+
+      if (!media) return false;
+
+      // TODO: Would seem to leave room for issue if this is called
+      // after a video has played but is currently paused.
+      // Could possibly check for a duration and a positive current time.
+      return !media.paused;
+    },
+    mediaEvents: ['playing', 'emptied']
+  },
+  MEDIA_PLAYBACK_RATE: {
+    get: function(controller) {
+      const { media } = controller;
 
-  #hotKeys = new AttributeTokenList(this, 'hotkeys');
-  #fullscreenElement;
+      if (!media || typeof media.playbackRate == 'undefined') {
+        return 1;
+      }
 
-  constructor() {
-    super();
+      return media.playbackRate;
+    },
+    mediaEvents: ['ratechange','loadstart']
+  },
+  MEDIA_MUTED: {
+    get: function(controller) {
+      const { media } = controller;
+
+      if (!media || typeof media.muted == 'undefined') {
+        return false;
+      }
+
+      return media.muted;
+    },
+    mediaEvents: ['volumechange']
+  },
+  MEDIA_VOLUME: {
+    get: function(controller) {
+      const { media } = controller;
+
+      if (!media || typeof media.volume == 'undefined') {
+        return 1;
+      }
+
+      return Number(media.volume);
+    },
+    mediaEvents: ['volumechange']
+  },
+  MEDIA_VOLUME_LEVEL: {
+    get: function(controller) {
+      const { media } = controller;
+      let level = 'high';
+
+      if (!media || typeof media.volume == 'undefined') {
+        return level;
+      }
+    
+      const { muted, volume } = media;
+    
+      if (volume === 0 || muted) {
+        level = 'off';
+      } else if (volume < 0.5) {
+        level = 'low';
+      } else if (volume < 0.75) {
+        level = 'medium';
+      }
+    
+      return level;
+    },
+    mediaEvents: ['volumechange']
+  },
+  MEDIA_CURRENT_TIME: {
+    get: function(controller) {
+      const { media } = controller;
+
+      if (!media || typeof media.currentTime == 'undefined') {
+        return 0;
+      }
+
+      return media.currentTime;
+    },
+    mediaEvents: ['timeupdate', 'loadedmetadata']
+  },
+  MEDIA_DURATION: {
+    get: function(controller){
+      const { media } = controller;
+
+      // TODO: What if duration is infinity/live? (heff)
+      if (!media || !Number.isFinite(media.duration)) {
```

Comment:
@cjpillsbury we shouldn't return NaN here when duration is Infinity (live), right?

Reply:
There are some assumptions in our time-related UI elements/Media State Receivers that I believe would need to be cleaned up. tl;dr - we prefer `seekable` but aren't guaranteed that the slotted media element supports `seekable`

Reply:
Got it. It feels odd to change what the media element is providing, but I guess it's fair to say at the UI level, if we're already using Infinity to set StreamType otherwise, then it simplifies UI logic to just return a falsey value here.

Reply:
I actually agree 💯. We can instead consider an "appropriate response/UI update" in the components themselves? I'm down for that.